File: 相助.cpp
#include <bits/stdc++.h>
using namespace std;

#define INF 0x3f3f3f3f

signed main() {
    int n;
    cin >> n;
    vector<int> dp(n + 1, INF), best(n + 1, INF);
    dp[0] = 0;
    for (int i = 1; i <= n; ++i) {
        int x;
        cin >> x;
        dp[i] = best[x] + 1;
        best[x] = min(best[x], dp[i - 1]);
    }
    cout << (dp[n] > n ? -1 : dp[n]) << endl;
    return 0;
}

File: 相助.py
n = int(input())
X = list(map(int, input().split()))
dp = [0x3f3f3f3f for _ in range(n + 1)]
best = [0x3f3f3f3f for _ in range(n + 1)]
dp[0] = 0
for i, x in enumerate(X, 1):
    dp[i] = best[x] + 1
    best[x] = min(best[x], dp[i - 1])
print(-1 if dp[n] > n else dp[n])
